### DISCUSSION TITLE
Devtools automatic logging

### DISCUSSION THREAD

**Discussion by lucasrabiec** [pmndrs/zustand#1185] [Ideas]
TL;DR
The idea is to have a flag (and logic) in devtools that will log the action's type based on store name combined with action name + payload (if exists). It should look like this: `{ type: appStateNameOrSliceName/actionName, payload: actionPayload }`

---

Regarding to: https://github.com/pmndrs/zustand/pull/1183 instead of adding `{ type: 'someAction', payload: 'actionPayload' }` to each action like this:

```ts
someUpdate: (somePayload) =>
  set(
    (state) => ..., // update state with payload
    false,
    { type: 'appStateNameOrSliceName/someUpdate', payload: somePayload },
  ),
```

It would be nice to have a flag in devtools options to enable automatic logging:

```ts
export interface DevtoolsOptions {
  automaticLogging: boolean
  ...rest
}
```

then action would look like this:
```ts
someUpdate: (somePayload) =>
  set(
    (state) => ..., // update state with payload
    # no need to add boolean and actionType
  ),
```
and it should log `{ type: appStateNameOrSliceName/someUpdate, payload: somePayload }` automatically in `Action` tab of Redux devtools.

**Comment by lucasrabiec** [comment]
When the developer adds his own action type it should replace that one that was added automatically by Zustand.

**Reply by dai-shi** [reply]
Zustand is designed to be predictable, no magic, no automatic things.
However, since this is a DEV-only feature, it might be okay.

I'm not exactly sure if it's easy to implement this without gotchas (and how easy it is to keep "maintaining" the feature.)
So, my suggestion for you (or someone) to work on a 3rd-party library. (and list it in https://github.com/pmndrs/zustand/blob/main/docs/3rd-party-libraries.md )

**Comment by devanshj** [comment]
You can make your own middleware for that btw, I think something like the following should work...

```ts
import create, { StateCreator, StoreMutatorIdentifier, StoreApi, Mutate } from 'zustand'
import { devtools } from "zustand/middleware"

type DevtoolsActionLogger =
  < T extends object
  , Mps extends [StoreMutatorIdentifier, unknown][] = []
  , Mcs extends [StoreMutatorIdentifier, unknown][] = []
  , U = T
  >
    (f:
      & StateCreator<T, Mps, Mcs, U>
      & StateCreator<T, [["zustand/devtools", never]], [StoreMutatorIdentifier, unknown][], U>
    , prefix?: string
    ) =>
      StateCreator<T, Mps, Mcs, U>

type DevtoolsActionLoggerImpl =
  <T extends object>
    (f: PopArgument<StateCreator<T, [], []>>, prefix?: string) =>
      PopArgument<StateCreator<T, [], []>>

const devtoolsActionLoggerImpl: DevtoolsActionLoggerImpl = (f, prefix = "") => (set, get, _store) => {
  type T = ReturnType<typeof f>
  let store = _store as Mutate<StoreApi<T>, [["zustand/devtools", never]]>

  let lastAction: string | ({ type: unknown } & { [_ in string]: unknown });

  let originalSetState = store.setState
  store.setState = (...a) => {
    if (a[2] !== undefined) {
      originalSetState(...a)
      return
    }
    originalSetState(a[0], a[1], lastAction)
  }
  
  let state = f(store.setState, get, store)
  
  return Object.fromEntries(Object.entries(state).map(([key, value]) =>
    typeof value !== "function" ? [key, value] :
    [key, (...a: unknown[]) => {
      lastAction = { type: prefix + key, arguments: a }
      return value(...a)
    }]
  )) as T
}

export const devtoolsActionLogger = devtoolsActionLoggerImpl as unknown as DevtoolsActionLogger

type PopArgument<T extends (...a: never[]) => unknown> =
  T extends (...a: [...infer A, infer _]) => infer R
    ? (...a: A) => R
    : never

// ---

interface BearSlice {
  bears: number
  addBear: () => void
  eatFish: () => void
}
const createBearSlice: StateCreator<
  BearSlice & FishSlice,
  [["zustand/devtools", BearSlice & FishSlice]],
  [],
  BearSlice
> = (set) => ({
  bears: 0,
  addBear: () => set((state) => ({ bears: state.bears + 1 })),
  eatFish: () => set((state) => ({ fishes: state.fishes - 1 })),
})

interface FishSlice {
  fishes: number
  addFish: () => void
}
const createFishSlice: StateCreator<
  BearSlice & FishSlice,
  [["zustand/devtools", BearSlice & FishSlice]],
  [],
  FishSlice
> = (set) => ({
  fishes: 0,
  addFish: () => set((state) => ({ fishes: state.fishes + 1 })),
})

const useRootStore = create<BearSlice & FishSlice>()(devtools((...a) => ({
  ...devtoolsActionLogger(createBearSlice, "bear/")(...a),
  ...devtoolsActionLogger(createFishSlice, "fish/")(...a),
})))
```
